### PR TITLE
Remove Auth Domain on Channel Delete

### LINKED ISSUE #2676: `hosting:channel:delete` doesn't remove authorized domain
### [REQUIRED] Environment info

**firebase-tools:** 8.12.0

**Platform:** Linux

### [REQUIRED] Test case

When creating/deploying a hosting channel, firebase requires the `Firebase Authentication Admin` role in order to allow the channel domain. However, when deleting a hosting channel the domain is not cleared up and is kept dangling.

### [REQUIRED] Steps to reproduce

1. From a firebase hosting project run: `firebase hosting:channel:create foo`;
2. `appname--foo-5myhd9tb.web.app` is added to the list of authorized domains;
3. Run `firebase hosting:channel:delete -f foo`;
4. `appname--foo-5myhd9tb.web.app` still listed as an authorized domain;

### [REQUIRED] Expected behavior

Running `firebase hosting:channel:delete -f foo`, should also delete the channel domain from the authorized domain list;

### [REQUIRED] Actual behavior

The domain is kept dangling in the list or authorized domains

### Logs
<details>
<summary><code>firebase hosting:channel:create --debug foo</code></summary>

```markdown
[2020-10-07T09:48:36.361Z] > command requires scopes: ["email","openid","https://www.googleapis.com/auth/cloudplatformprojects.readonly","https://www.googleapis.com/auth/firebase","https://www.googleapis.com/auth/cloud-platform"]
[2020-10-07T09:48:36.362Z] > authorizing via FIREBASE_TOKEN environment variable
[2020-10-07T09:48:36.362Z] [iam] checking project appdev for permissions ["firebase.projects.get","firebasehosting.sites.update"]
[2020-10-07T09:48:36.363Z] > refreshing access token with scopes: ["email","openid","https://www.googleapis.com/auth/cloudplatformprojects.readonly","https://www.googleapis.com/auth/firebase","https://www.googleapis.com/auth/cloud-platform"]
[2020-10-07T09:48:36.363Z] >>> HTTP REQUEST POST https://www.googleapis.com/oauth2/v3/token  
 <request body omitted>
[2020-10-07T09:48:36.520Z] <<< HTTP RESPONSE 200 {"content-type":"application/json; charset=utf-8","vary":"X-Origin, Referer, Origin,Accept-Encoding","date":"Wed, 07 Oct 2020 09:48:54 GMT","server":"scaffolding on HTTPServer2","cache-control":"private","x-xss-protection":"0","x-frame-options":"SAMEORIGIN","x-content-type-options":"nosniff","alt-svc":"h3-Q050=\":443\"; ma=2592000,h3-29=\":443\"; ma=2592000,h3-27=\":443\"; ma=2592000,h3-T051=\":443\"; ma=2592000,h3-T050=\":443\"; ma=2592000,h3-Q046=\":443\"; ma=2592000,h3-Q043=\":443\"; ma=2592000,quic=\":443\"; ma=2592000; v=\"46,43\"","accept-ranges":"none","transfer-encoding":"chunked"}
[2020-10-07T09:48:36.525Z] >>> HTTP REQUEST POST https://cloudresourcemanager.googleapis.com/v1/projects/appdev:testIamPermissions  
 {"permissions":["firebase.projects.get","firebasehosting.sites.update"]}
[2020-10-07T09:48:37.270Z] <<< HTTP RESPONSE 200 {"content-type":"application/json; charset=UTF-8","vary":"X-Origin, Referer, Origin,Accept-Encoding","date":"Wed, 07 Oct 2020 09:48:55 GMT","server":"ESF","cache-control":"private","x-xss-protection":"0","x-frame-options":"SAMEORIGIN","x-content-type-options":"nosniff","server-timing":"gfet4t7; dur=655","alt-svc":"h3-Q050=\":443\"; ma=2592000,h3-29=\":443\"; ma=2592000,h3-27=\":443\"; ma=2592000,h3-T051=\":443\"; ma=2592000,h3-T050=\":443\"; ma=2592000,h3-Q046=\":443\"; ma=2592000,h3-Q043=\":443\"; ma=2592000,quic=\":443\"; ma=2592000; v=\"46,43\"","accept-ranges":"none","transfer-encoding":"chunked"}
[2020-10-07T09:48:37.271Z] >>> HTTP REQUEST GET https://firebase.googleapis.com/v1beta1/projects/appdev  
 
[2020-10-07T09:48:37.581Z] <<< HTTP RESPONSE 200 {"content-type":"application/json; charset=UTF-8","vary":"X-Origin, Referer, Origin,Accept-Encoding","date":"Wed, 07 Oct 2020 09:48:55 GMT","server":"ESF","cache-control":"private","x-xss-protection":"0","x-frame-options":"SAMEORIGIN","x-content-type-options":"nosniff","alt-svc":"h3-Q050=\":443\"; ma=2592000,h3-29=\":443\"; ma=2592000,h3-27=\":443\"; ma=2592000,h3-T051=\":443\"; ma=2592000,h3-T050=\":443\"; ma=2592000,h3-Q046=\":443\"; ma=2592000,h3-Q043=\":443\"; ma=2592000,quic=\":443\"; ma=2592000; v=\"46,43\"","accept-ranges":"none","transfer-encoding":"chunked"}
[2020-10-07T09:48:37.585Z] >>> HTTP REQUEST GET https://firebase.googleapis.com/v1beta1/projects/appdev  
 
[2020-10-07T09:48:37.794Z] <<< HTTP RESPONSE 200 {"content-type":"application/json; charset=UTF-8","vary":"X-Origin, Referer, Origin,Accept-Encoding","date":"Wed, 07 Oct 2020 09:48:56 GMT","server":"ESF","cache-control":"private","x-xss-protection":"0","x-frame-options":"SAMEORIGIN","x-content-type-options":"nosniff","alt-svc":"h3-Q050=\":443\"; ma=2592000,h3-29=\":443\"; ma=2592000,h3-27=\":443\"; ma=2592000,h3-T051=\":443\"; ma=2592000,h3-T050=\":443\"; ma=2592000,h3-Q046=\":443\"; ma=2592000,h3-Q043=\":443\"; ma=2592000,quic=\":443\"; ma=2592000; v=\"46,43\"","accept-ranges":"none","transfer-encoding":"chunked"}
[2020-10-07T09:48:37.796Z] >>> HTTP REQUEST POST https://firebasehosting.googleapis.com/v1beta1/projects/appdev/sites/appdev/channels?channelId=foo  
 {"ttl":"604800s"}
[2020-10-07T09:48:38.754Z] <<< HTTP RESPONSE 200 {"content-type":"application/json; charset=UTF-8","vary":"X-Origin, Referer, Origin,Accept-Encoding","date":"Wed, 07 Oct 2020 09:48:56 GMT","server":"ESF","cache-control":"private","x-xss-protection":"0","x-frame-options":"SAMEORIGIN","x-content-type-options":"nosniff","alt-svc":"h3-Q050=\":443\"; ma=2592000,h3-29=\":443\"; ma=2592000,h3-27=\":443\"; ma=2592000,h3-T051=\":443\"; ma=2592000,h3-T050=\":443\"; ma=2592000,h3-Q046=\":443\"; ma=2592000,h3-Q043=\":443\"; ma=2592000,quic=\":443\"; ma=2592000; v=\"46,43\"","accept-ranges":"none","transfer-encoding":"chunked"}
[2020-10-07T09:48:38.757Z] >>> HTTP REQUEST GET https://identitytoolkit.googleapis.com/admin/v2/projects/appdev/config  
 
[2020-10-07T09:48:39.175Z] <<< HTTP RESPONSE 200 {"date":"Wed, 07 Oct 2020 09:48:57 GMT","pragma":"no-cache","expires":"Mon, 01 Jan 1990 00:00:00 GMT","cache-control":"no-cache, no-store, max-age=0, must-revalidate","content-type":"application/json; charset=UTF-8","vary":"X-Origin, Referer, Origin,Accept-Encoding","server":"ESF","x-xss-protection":"0","x-frame-options":"SAMEORIGIN","x-content-type-options":"nosniff","alt-svc":"h3-Q050=\":443\"; ma=2592000,h3-29=\":443\"; ma=2592000,h3-27=\":443\"; ma=2592000,h3-T051=\":443\"; ma=2592000,h3-T050=\":443\"; ma=2592000,h3-Q046=\":443\"; ma=2592000,h3-Q043=\":443\"; ma=2592000,quic=\":443\"; ma=2592000; v=\"46,43\"","accept-ranges":"none","transfer-encoding":"chunked"}

# THIS REQUEST SHOULD BE REVERSED ON DELETION
[2020-10-07T09:48:39.176Z] >>> HTTP REQUEST PATCH https://identitytoolkit.googleapis.com/admin/v2/projects/appdev/config?update_mask=authorizedDomains  
 {"authorizedDomains":["localhost","appdev.firebaseapp.com","appdev.web.app","appdev--foobar-upbqc8zc.web.app","appdev--foo-5myhd9tb.web.app","appdev--foo-amn3rax9.web.app","appdev--foo-dkdqxg4h.web.app"]}
[2020-10-07T09:48:39.507Z] <<< HTTP RESPONSE 200 {"cache-control":"no-cache, no-store, max-age=0, must-revalidate","date":"Wed, 07 Oct 2020 09:48:57 GMT","pragma":"no-cache","expires":"Mon, 01 Jan 1990 00:00:00 GMT","content-type":"application/json; charset=UTF-8","vary":"X-Origin, Referer, Origin,Accept-Encoding","server":"ESF","x-xss-protection":"0","x-frame-options":"SAMEORIGIN","x-content-type-options":"nosniff","alt-svc":"h3-Q050=\":443\"; ma=2592000,h3-29=\":443\"; ma=2592000,h3-27=\":443\"; ma=2592000,h3-T051=\":443\"; ma=2592000,h3-T050=\":443\"; ma=2592000,h3-Q046=\":443\"; ma=2592000,h3-Q043=\":443\"; ma=2592000,quic=\":443\"; ma=2592000; v=\"46,43\"","accept-ranges":"none","transfer-encoding":"chunked"}

✔  hosting:channel: Channel foo has been created on site appdev. 
✔  hosting:channel: Channel foo will expire at 2020-10-14 09:48:56. 
✔  hosting:channel: Channel URL: https://appdev--foo-dkdqxg4h.web.app 

To deploy to this channel, use `firebase hosting:channel:deploy foo`.
```
</details>

<details>
<summary><code>firebase hosting:channel:delete -f --debug foo</code></summary>

```markdown
[2020-10-07T09:48:55.932Z] > command requires scopes: ["email","openid","https://www.googleapis.com/auth/cloudplatformprojects.readonly","https://www.googleapis.com/auth/firebase","https://www.googleapis.com/auth/cloud-platform"]
[2020-10-07T09:48:55.933Z] > authorizing via FIREBASE_TOKEN environment variable
[2020-10-07T09:48:55.933Z] [iam] checking project appdev for permissions ["firebase.projects.get","firebasehosting.sites.update"]
[2020-10-07T09:48:55.935Z] > refreshing access token with scopes: ["email","openid","https://www.googleapis.com/auth/cloudplatformprojects.readonly","https://www.googleapis.com/auth/firebase","https://www.googleapis.com/auth/cloud-platform"]
[2020-10-07T09:48:55.935Z] >>> HTTP REQUEST POST https://www.googleapis.com/oauth2/v3/token  
 <request body omitted>
[2020-10-07T09:48:56.304Z] <<< HTTP RESPONSE 200 {"content-type":"application/json; charset=utf-8","vary":"X-Origin, Referer, Origin,Accept-Encoding","date":"Wed, 07 Oct 2020 09:49:14 GMT","server":"scaffolding on HTTPServer2","cache-control":"private","x-xss-protection":"0","x-frame-options":"SAMEORIGIN","x-content-type-options":"nosniff","alt-svc":"h3-Q050=\":443\"; ma=2592000,h3-29=\":443\"; ma=2592000,h3-27=\":443\"; ma=2592000,h3-T051=\":443\"; ma=2592000,h3-T050=\":443\"; ma=2592000,h3-Q046=\":443\"; ma=2592000,h3-Q043=\":443\"; ma=2592000,quic=\":443\"; ma=2592000; v=\"46,43\"","accept-ranges":"none","transfer-encoding":"chunked"}
[2020-10-07T09:48:56.308Z] >>> HTTP REQUEST POST https://cloudresourcemanager.googleapis.com/v1/projects/appdev:testIamPermissions  
 {"permissions":["firebase.projects.get","firebasehosting.sites.update"]}
[2020-10-07T09:48:57.173Z] <<< HTTP RESPONSE 200 {"content-type":"application/json; charset=UTF-8","vary":"X-Origin, Referer, Origin,Accept-Encoding","date":"Wed, 07 Oct 2020 09:49:15 GMT","server":"ESF","cache-control":"private","x-xss-protection":"0","x-frame-options":"SAMEORIGIN","x-content-type-options":"nosniff","server-timing":"gfet4t7; dur=653","alt-svc":"h3-Q050=\":443\"; ma=2592000,h3-29=\":443\"; ma=2592000,h3-27=\":443\"; ma=2592000,h3-T051=\":443\"; ma=2592000,h3-T050=\":443\"; ma=2592000,h3-Q046=\":443\"; ma=2592000,h3-Q043=\":443\"; ma=2592000,quic=\":443\"; ma=2592000; v=\"46,43\"","accept-ranges":"none","transfer-encoding":"chunked"}
[2020-10-07T09:48:57.174Z] >>> HTTP REQUEST GET https://firebase.googleapis.com/v1beta1/projects/appdev  
 
[2020-10-07T09:48:57.533Z] <<< HTTP RESPONSE 200 {"content-type":"application/json; charset=UTF-8","vary":"X-Origin, Referer, Origin,Accept-Encoding","date":"Wed, 07 Oct 2020 09:49:15 GMT","server":"ESF","cache-control":"private","x-xss-protection":"0","x-frame-options":"SAMEORIGIN","x-content-type-options":"nosniff","alt-svc":"h3-Q050=\":443\"; ma=2592000,h3-29=\":443\"; ma=2592000,h3-27=\":443\"; ma=2592000,h3-T051=\":443\"; ma=2592000,h3-T050=\":443\"; ma=2592000,h3-Q046=\":443\"; ma=2592000,h3-Q043=\":443\"; ma=2592000,quic=\":443\"; ma=2592000; v=\"46,43\"","accept-ranges":"none","transfer-encoding":"chunked"}
[2020-10-07T09:48:57.534Z] >>> HTTP REQUEST GET https://firebase.googleapis.com/v1beta1/projects/appdev  
 
[2020-10-07T09:48:57.893Z] <<< HTTP RESPONSE 200 {"content-type":"application/json; charset=UTF-8","vary":"X-Origin, Referer, Origin,Accept-Encoding","date":"Wed, 07 Oct 2020 09:49:16 GMT","server":"ESF","cache-control":"private","x-xss-protection":"0","x-frame-options":"SAMEORIGIN","x-content-type-options":"nosniff","alt-svc":"h3-Q050=\":443\"; ma=2592000,h3-29=\":443\"; ma=2592000,h3-27=\":443\"; ma=2592000,h3-T051=\":443\"; ma=2592000,h3-T050=\":443\"; ma=2592000,h3-Q046=\":443\"; ma=2592000,h3-Q043=\":443\"; ma=2592000,quic=\":443\"; ma=2592000; v=\"46,43\"","accept-ranges":"none","transfer-encoding":"chunked"}
[2020-10-07T09:48:57.896Z] >>> HTTP REQUEST DELETE https://firebasehosting.googleapis.com/v1beta1/projects/appdev/sites/appdev/channels/foo  
 
[2020-10-07T09:48:59.051Z] <<< HTTP RESPONSE 200 {"content-type":"application/json; charset=UTF-8","vary":"X-Origin, Referer, Origin,Accept-Encoding","date":"Wed, 07 Oct 2020 09:49:17 GMT","server":"ESF","cache-control":"private","x-xss-protection":"0","x-frame-options":"SAMEORIGIN","x-content-type-options":"nosniff","alt-svc":"h3-Q050=\":443\"; ma=2592000,h3-29=\":443\"; ma=2592000,h3-27=\":443\"; ma=2592000,h3-T051=\":443\"; ma=2592000,h3-T050=\":443\"; ma=2592000,h3-Q046=\":443\"; ma=2592000,h3-Q043=\":443\"; ma=2592000,quic=\":443\"; ma=2592000; v=\"46,43\"","accept-ranges":"none","transfer-encoding":"chunked"}
✔  hosting:channels: Successfully deleted channel foo for site appdev. 
```
</details>

### PR DESCRIPTION
**Pull request by dahliasalem**
### Description

Removes the channel from the authorized domains when the channel is deleted.

Fixes #2676.

### CODE REVIEW

File: src/hosting/api.ts [hunk 1]
```diff
@@ -373,6 +373,27 @@ export async function addAuthDomain(project: string, url: string): Promise<strin
...
 }
 
+/**
+ * Removes channel domain from Firebase Auth list.
+ * @param project the project ID.
+ * @param url the url of the channel.
+ */
+export async function removeAuthDomain(project: string, url: string): Promise<string[]> {
+  const domains = await getAuthDomains(project);
+  if (domains.length < 1) {
```
Comment (bkendall) [inline src/hosting/api.ts:383]:
```suggestion
  if (!domains.length) {
```

File: src/hosting/api.ts [hunk 2]
```diff
@@ -373,6 +373,27 @@ export async function addAuthDomain(project: string, url: string): Promise<strin
...
+    return domains;
+  }
+  const targetDomain = url.replace("https://", "");
+  const authDomains: string[] = [];
+  domains.forEach((domain: string) => {
+    if (domain == targetDomain) {
+      return;
+    }
+    authDomains.push(domain);
+  });
```
Comment (bkendall) [inline src/hosting/api.ts:393]:
```typescript
const authDomains = domains.forEach((d: string) => d != targetDomain);
```

(untested, but I believe that passes lint rules)

**Review by bkendall — CHANGES_REQUESTED** [review]
Couple tidbits to clean up, but LGTM.

Please add a changelog entry:

```typescript
- Updates `firebase hosting:channel:delete` to remove the channel from the authorized domains list.
```

File: src/hosting/api.ts [hunk 3]
```diff
@@ -373,6 +373,27 @@ export async function addAuthDomain(project: string, url: string): Promise<strin
...
+  return await updateAuthDomains(project, authDomains);
```
Comment (bkendall) [inline src/hosting/api.ts:394]: Isn't `return updateAuthDomains` sufficient? Can `await` be omitted here?

File: src/hosting/api.ts (hunk 2 above)
Comment (dahliasalem) [reply src/hosting/api.ts:393]: I think you meant `.filter`

**Review by bkendall — APPROVED** [review]

**Pull request merged**